pkg/ddevapp/utils.go

```diff
@@ -400,6 +401,44 @@ func ExtractProjectNames(apps []*DdevApp) []string {
 	return names
 }
 
+// GetProjectNamesFunc returns a function for autocompleting project names
+// for command arguments.
+// If status is "inactive" or "active", only names of inactive or active
+// projects respectively are returned.
+// If status is "all", all project names are returned.
+// If numArgs is 0, completion will be provided for infinite arguments,
+// otherwise it will only be provided for the numArgs number of arguments.
+func GetProjectNamesFunc(status string, numArgs int) func(*cobra.Command, []string, string) ([]string, cobra.ShellCompDirective) {
+	return func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		// Don't provide completions if the user keeps hitting space after
+		// exhausting all of the valid arguments.
+		if numArgs > 0 && len(args)+1 > numArgs {
+			return nil, cobra.ShellCompDirectiveNoFileComp
+		}
+
+		// Get all of the projects we're interested in for this completion function.
+		var apps []*DdevApp
+		var err error
+		if status == "inactive" {
+			apps, err = GetInactiveProjects()
+		} else if status == "active" {
+			apps, err = GetProjects(true)
+		} else if status == "all" {
+			apps, err = GetProjects(false)
+		} else {
+			// This is an error state - but we just return nothing
+			return nil, cobra.ShellCompDirectiveNoFileComp
+		}
+		// Return nothing if we have nothing, or return all of the project names.
+		// Note that if there's nothing to return, we don't let cobra pick completions
+		// from the files in the cwd.
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveNoFileComp
+		}
+		return ExtractProjectNames(apps), cobra.ShellCompDirectiveNoFileComp
```

Comment:
```suggestion
		// Do not show arguments that are already written on the command line
		var appNames []string
		appNamesFromCmd := make(map[string]struct{})
		for _, name := range cmd.Flags().Args() {
			appNamesFromCmd[name] = struct{}{}
		}
		for _, name := range ExtractProjectNames(apps) {
			if _, exists := appNamesFromCmd[name]; !exists {
				appNames = append(appNames, name)
			}
		}
		return appNames, cobra.ShellCompDirectiveNoFileComp
```

Reply:
The suggestion is a fix for:

- `ddev start project-name <TAB>` and `project-name` is still in the list.
- one project is running and you run `ddev stop <TAB><TAB><TAB>`, the result is `ddev stop project-name project-name project-name`.
- and so on.

Reply:
Ooh good catch.

Reply:
I did something a little different than the suggestion, but the result is the same.
- Called the variable `projectNames` instead of `appNames` since that's the terminology used in the rest of the function
- Used the `args` argument that cobra is giving us instead of fetching the args manually.

Reply:
Thanks, that looks better 👍, because I just googled how to do `array_intersect()` in go and used the first option I found.